File: experiment/kernel.py
"""Structured product kernel (manuscript Section 4).

    k(x, x') = sigma^2 * k_tool(a, a') * k_ctx(c, c') * k_time(t, t')

* ``k_tool``  RBF over the tool feature block (one-hot name + semantic
  descriptor): shared tool name and shared reversibility/category raise
  similarity, exactly the manuscript's "shared tool name, overlapping
  argument patterns, same reversibility class".
* ``k_ctx``   RBF over the context block (target sensitivity, destructive-arg
  flag, task one-hot): "same repository, file type, task category".
* ``k_time``  ``exp(-|t - t'| / lambda)``, the manuscript's Section 6 kernel
  for non-stationarity. This is the Ornstein-Uhlenbeck covariance and is
  positive semidefinite; the product of PSD kernels is PSD (Schur product).

A single signal variance ``sigma^2`` multiplies the product (per-block scales
are not separately identifiable). Each block's self-similarity is 1, so the
prior variance on the diagonal is exactly ``sigma^2``.
"""

from __future__ import annotations

from dataclasses import dataclass

import numpy as np

from .data import TOOL_INDEX, DecisionPoint


@dataclass
class Packed:
    """Stream packed into dense arrays for vectorized kernel evaluation.

    ``tool_id`` is carried for the independent baseline only; the kernel never
    uses it (tool similarity comes from the semantic block of ``phi_tool``).
    """

    phi_tool: np.ndarray  # (N, d_tool)
    phi_ctx: np.ndarray   # (N, d_ctx)
    t: np.ndarray         # (N,)
    tool_id: np.ndarray   # (N,)


def pack(stream: list[DecisionPoint]) -> Packed:
    return Packed(
        phi_tool=np.stack([dp.phi_tool for dp in stream]),
        phi_ctx=np.stack([dp.phi_ctx for dp in stream]),
        t=np.array([dp.t for dp in stream], dtype=float),
        tool_id=np.array([TOOL_INDEX[dp.tool.name] for dp in stream]),
    )


def _sqdist(A: np.ndarray, B: np.ndarray) -> np.ndarray:
    """Pairwise squared Euclidean distance between rows of A and B."""
    a2 = np.sum(A * A, axis=1)[:, None]
    b2 = np.sum(B * B, axis=1)[None, :]
    d2 = a2 + b2 - 2.0 * A @ B.T
    return np.maximum(d2, 0.0)


@dataclass
class ProductKernel:
    """Manuscript Section 4 product kernel with structured blocks."""

    sigma2: float = 1.6
    l_tool: float = 1.1
    l_ctx: float = 1.2
    lam: float = 200.0  # time lengthscale (steps); ~ the drift timescale

    def _blocks(self, P: Packed, Q: Packed) -> np.ndarray:
        k_tool = np.exp(-_sqdist(P.phi_tool, Q.phi_tool) / (2.0 * self.l_tool**2))
        k_ctx = np.exp(-_sqdist(P.phi_ctx, Q.phi_ctx) / (2.0 * self.l_ctx**2))
        dt = np.abs(P.t[:, None] - Q.t[None, :])
        k_time = np.exp(-dt / self.lam)
        return self.sigma2 * k_tool * k_ctx * k_time

    def full(self, P: Packed) -> np.ndarray:
        """Train-train covariance ``K``."""
        return self._blocks(P, P)

    def cross(self, P: Packed, Q: Packed) -> np.ndarray:
        """Train-test cross covariance ``K_*`` (rows P, cols Q)."""
        return self._blocks(P, Q)

    def diag(self, P: Packed) -> np.ndarray:
        """Prior variances ``diag(K)`` (each block self-similarity is 1)."""
        return np.full(P.phi_tool.shape[0], self.sigma2, dtype=float)

File: go/kernel/kernel.go
// Package kernel implements the structured product kernel from Section 4 of
// the trustcalib manuscript:
//
//	k(x, x') = sigma2 * k_tool(a, a') * k_ctx(c, c') * k_time(t, t')
//
// k_tool and k_ctx are squared-exponential (RBF) kernels over the tool and
// context feature blocks; k_time = exp(-|t-t'|/lambda) is the Ornstein-
// Uhlenbeck covariance for non-stationarity. The product of PSD kernels is PSD
// (Schur product). Each block's self-similarity is 1, so the prior variance on
// the diagonal is exactly sigma2.
//
// The kernel is stateless and operates on raw feature vectors (a Packed),
// independent of any domain taxonomy.
package kernel

import (
	"math"

	"gonum.org/v1/gonum/mat"
)

// Packed is a batch of decision points packed into dense arrays for vectorized
// kernel evaluation. PhiTool is (N x dTool), PhiCtx is (N x dCtx) and T has
// length N.
type Packed struct {
	PhiTool *mat.Dense
	PhiCtx  *mat.Dense
	T       []float64
}

// Len returns the number of points.
func (p Packed) Len() int { return len(p.T) }

// ProductKernel holds the kernel hyperparameters. The zero value is not valid;
// use DefaultKernel or set every field.
type ProductKernel struct {
	Sigma2 float64 // signal variance (prior diagonal)
	LTool  float64 // RBF lengthscale for the tool block
	LCtx   float64 // RBF lengthscale for the context block
	Lambda float64 // time lengthscale (OU decay, in steps)
}

// DefaultKernel returns the manuscript defaults (kernel.py dataclass defaults).
func DefaultKernel() ProductKernel {
	return ProductKernel{Sigma2: 1.6, LTool: 1.1, LCtx: 1.2, Lambda: 200.0}
}

// sqdist returns the (m x n) matrix of pairwise squared Euclidean distances
// between rows of a (m x d) and b (n x d), floored at 0 to clip numerical
// noise. Mirrors data.py / kernel.py _sqdist: a2 + b2 - 2 A Bᵀ.
func sqdist(a, b *mat.Dense) *mat.Dense {
	m, _ := a.Dims()
	n, _ := b.Dims()

	var cross mat.Dense
	cross.Mul(a, b.T())

	a2 := rowSqNorms(a, m)
	b2 := rowSqNorms(b, n)

	out := mat.NewDense(m, n, nil)
	for i := 0; i < m; i++ {
		for j := 0; j < n; j++ {
			v := a2[i] + b2[j] - 2.0*cross.At(i, j)
			if v < 0 {
				v = 0
			}
			out.Set(i, j, v)
		}
	}
	return out
}

func rowSqNorms(a *mat.Dense, m int) []float64 {
	out := make([]float64, m)
	for i := 0; i < m; i++ {
		s := 0.0
		for _, v := range a.RawRowView(i) {
			s += v * v
		}
		out[i] = s
	}
	return out
}

// blocks computes the full (m x n) product-kernel matrix between p and q.
func (k ProductKernel) blocks(p, q Packed) *mat.Dense {
	d2tool := sqdist(p.PhiTool, q.PhiTool)
	d2ctx := sqdist(p.PhiCtx, q.PhiCtx)
	m, n := d2tool.Dims()

	twoLTool2 := 2.0 * k.LTool * k.LTool
	twoLCtx2 := 2.0 * k.LCtx * k.LCtx

	out := mat.NewDense(m, n, nil)
	for i := 0; i < m; i++ {
		for j := 0; j < n; j++ {
			kTool := math.Exp(-d2tool.At(i, j) / twoLTool2)
			kCtx := math.Exp(-d2ctx.At(i, j) / twoLCtx2)
			kTime := math.Exp(-math.Abs(p.T[i]-q.T[j]) / k.Lambda)
			out.Set(i, j, k.Sigma2*kTool*kCtx*kTime)
		}
	}
	return out
}

// Full returns the symmetric (N x N) train-train covariance K.
func (k ProductKernel) Full(p Packed) *mat.SymDense {
	n := p.Len()
	b := k.blocks(p, p)
	s := mat.NewSymDense(n, nil)
	for i := 0; i < n; i++ {
		for j := i; j < n; j++ {
			s.SetSym(i, j, b.At(i, j))
		}
	}
	return s
}

// Cross returns the (len(p) x len(q)) train-test cross covariance (rows p,
// cols q).
func (k ProductKernel) Cross(p, q Packed) *mat.Dense {
	return k.blocks(p, q)
}

// Diag returns the prior variances diag(K); every entry equals Sigma2.
func (k ProductKernel) Diag(p Packed) []float64 {
	out := make([]float64, p.Len())
	for i := range out {
		out[i] = k.Sigma2
	}
	return out
}
